### PR TITLE
[NP-3587] Refactor CIDR Support

### PR DESCRIPTION
**Pull request by jlhughes**
* Refactor CIDR Support
* Improve CIDR Notation entry, showing resulting IP range.
* Add CIDR support Sessions - replace explicit IP address whitelist.

Screen Shot 2021-02-12 at 8 19 43 AM
Screen Shot 2021-02-12 at 8 18 49 AM

### CODE REVIEW

File: src/foam/nanos/session/Session.js [hunk 1]
```diff
@@ -185,25 +189,31 @@ foam.CLASS({
...
+          if ( cidr.inRange(x, remoteIp) ) {
+            return;
           }
+        } catch (java.net.UnknownHostException e) {
+          ((foam.nanos.logger.Logger) x.get("logger")).warning(this.getClass().getSimpleName(), "validateRemoteHost", remoteIp, e.getMessage());
         }
-
-        return false;
+      }
+      throw new foam.core.ValidationException("Restricted IP");
```
Comment (lchanmann) [inline src/foam/nanos/session/Session.js:219]: Getting "Access denied" when login as admin@nanopay.net.
```
2021-02-12T07:08:38.421-0500,qtp1189257362-1848,DEBUG,[Service],auth,Unauthenticated session,returning skeleton exception,[localhost],foam.nanos.auth.AccessDeniedException,Access denied
	at foam.nanos.auth.UserAndGroupAuthService.loginHelper(UserAndGroupAuthService.java:335)
	at foam.nanos.auth.UserAndGroupAuthService.login(UserAndGroupAuthService.java:390)
	at net.nanopay.auth.NanopayUserAndGroupAuthService.login_(NanopayUserAndGroupAuthService.java:228)
	at net.nanopay.auth.NanopayUserAndGroupAuthService.login(NanopayUserAndGroupAuthService.java:191)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at foam.nanos.auth.PasswordExpiryAuthService.login(PasswordExpiryAuthService.java:120)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at net.nanopay.security.auth.LoginAttemptAuthService.login_(LoginAttemptAuthService.java:546)
	at net.nanopay.security.auth.LoginAttemptAuthService.login(LoginAttemptAuthService.java:503)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at net.nanopay.security.auth.IPLoggingAuthService.login(IPLoggingAuthService.java:92)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at foam.nanos.auth.AuthServiceSkeleton.send(AuthServiceSkeleton.java:104)
	at foam.box.SessionServerBox.send(SessionServerBox.java:185)
	at foam.nanos.http.ServiceWebAgent.execute(ServiceWebAgent.java:129)
	at foam.nanos.http.NanoRouter.service(NanoRouter.java:127)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:763)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1631)
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:226)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at net.nanopay.security.csp.CSPFilter.doFilter(CSPFilter.java:24)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:549)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1369)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:489)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1284)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:501)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel$$Lambda$100/0000000000000000.dispatch(Unknown Source)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:272)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: [localhost],foam.core.ValidationException,Restricted IP
	at foam.nanos.auth.Group.validateCidrWhiteList(Group.java:1426)
	at foam.nanos.auth.UserAndGroupAuthService.loginHelper(UserAndGroupAuthService.java:333)
	... 52 more
```

File: src/foam/net/CIDR.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+          return;
+        }
+        var addr32 = m.slice(0, 4).reduce(function (a, o) {
+          return u(+a << 8) + +o;
+        });                           // 0xc6a20101
+        let mask = u(~0 << (32 - +m[4])); // 0xffffff00
+        this.low = u(addr32 & mask);
+        this.high = u(addr32 | ~mask);
+        this.startAddress = ip(this.low);
+        this.endAddress = ip(this.high);
```
Comment (lchanmann) [inline src/foam/net/CIDR.js:55]: The calculated range includes network and broadcast addresses, which are not assignable. Do we want to include them?

File: src/foam/net/CIDR.js [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+      }
+      long high = getHigh();
+      if ( high == 0 ) {
+        high = ip2Long(getEndAddress());
+        setHigh(high);
+      }
+      long test = ip2Long(ip);
+
+      return test >= low &&
+             test <= high;
```
Comment (lchanmann) [inline src/foam/net/CIDR.js:113]: This depends on low, high or start, end addresses properties but only the JS implementation exists for the calculation.

File: src/foam/net/IPSupport.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      javaCode: `
+      HttpServletRequest req = x.get(HttpServletRequest.class);
+      if ( req == null ) {
+        return null;
+      }
+      String forwardedForHeader = req.getHeader("X-Forwarded-For");
+      if ( ! foam.util.SafetyUtil.isEmpty(forwardedForHeader) ) {
+        String[] addresses = forwardedForHeader.split(",");
+        return addresses[addresses.length -1].trim(); // right most
+      } else {
```
Comment (kgrgreer) [inline src/foam/net/IPSupport.js]: No need for 'else' when you've already returned in the 'if'.

**Comment by kgrgreer** [comment]
Could use more documentation to help explain what its for and why we're doing ti.

File: src/foam/net/CIDR.js (hunk 3 above)
Comment (jlhughes) [reply src/foam/net/CIDR.js:113]: Yes. you can explicitly set the high and low in config files if you like.

File: src/foam/net/CIDR.js (hunk 2 above)
Comment (jlhughes) [reply src/foam/net/CIDR.js:55]: This seems to be the norm.

File: src/foam/nanos/session/Session.js (hunk 1 above)
Comment (jlhughes) [reply src/foam/nanos/session/Session.js:219]: Are you saying you get this with no configuration?

Comment (jlhughes) [reply src/foam/nanos/session/Session.js:219]: Updated.

File: src/foam/net/CIDR.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.net',
+  name: 'CIDR',
+
+  documentation: `Model a Classless Inter-Domain Routing Range. See IPSupport for validating an IP against a list of CIDRs.  Used to restrict access to particular IP address range.
```
Comment (jlhughes) [inline src/foam/net/CIDR.js]:
```suggestion
  documentation: `Model a Classless Inter-Domain Routing Range.  Used to restrict access to particular IP address range.  So, for example, admin group users can only login when on a company VPN. 
```

**Review by nanoscott — APPROVED** [review]

**Pull request merged**